core/sail/shacl/src/test/java/org/eclipse/rdf4j/sail/shacl/AbstractShaclTest.java

```diff
@@ -538,6 +744,61 @@ static String modelToString(Model model) {
 		return stringWriter.toString();
 	}
 
+	private static Model extractValidationReport(Model model) {
+
+		Resource subject = Models.subject(model.filter(null, RDF.TYPE, SHACL.VALIDATION_REPORT)).get();
+		return ModelExtractor.extract(model, subject, s -> {
+			if (s.getPredicate().equals(SHACL.SOURCE_SHAPE))
+				return ModelExtractor.Decision.includeDontFollow;
+
+			return ModelExtractor.Decision.includeAndFollow;
+		});
+	}
+
+	static class ModelExtractor {
+
+		enum Decision {
+			includeAndFollow,
+			includeDontFollow,
+			exclude
+		}
+
+		static Model extract(Model model, Resource start, Function<Statement, Decision> decisionFunction) {
+			DynamicModel emptyModel = new DynamicModelFactory().createEmptyModel();
+
+			Set<Statement> breadthFirstSearchBuffer = new HashSet<>();
+			model.getStatements(start, null, null).forEach(breadthFirstSearchBuffer::add);
+
+			while (!breadthFirstSearchBuffer.isEmpty()) {
+				Set<Statement> tempBuffer = new HashSet<>();
+				for (Statement statement : breadthFirstSearchBuffer) {
+					Decision decision = decisionFunction.apply(statement);
+
+					switch (decision) {
+
+					case includeAndFollow:
+						boolean add = emptyModel.add(statement);
+						if (add && statement.getObject() instanceof Resource) {
+							model.getStatements((Resource) statement.getObject(), null, null).forEach(tempBuffer::add);
+						}
+						break;
+					case includeDontFollow:
+						emptyModel.add(statement);
+						break;
+					case exclude:
+						break;
+					}
+				}
+
+				breadthFirstSearchBuffer = tempBuffer;
+
+			}
+
+			return emptyModel;
+		}
+
+	}
+
```

Comment:
@jeenbroekstra do we have something like this already? It starts at a node and follows all outbound statements from that node using breadth first search. I use it to extract a subset of the validation report excluding the source shape.

If we don't have something like this I think it could be a useful addition to the Models helper class.

Reply:
Jena has something similar: https://jena.apache.org/documentation/javadoc/jena/org/apache/jena/graph/GraphExtract.html

Reply:
I don't think we have something quite like that right now. I've been thinking that more generally I'd like a few more graph-traversal kind of operations available in RDF4J, and this fits right in with that.

So yes, nice idea. Want to open a ticket for adding this as a new feature?